tests/Normalization.hs

```diff
@@ -1,18 +1,73 @@
-{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE OverloadedStrings#-}
 module Normalization (normalizationTests) where
 
 import           Dhall.Core
+import           Dhall.TypeCheck (typeOf, X(..))
 import           Test.Tasty
 import           Test.Tasty.HUnit
 
 normalizationTests :: TestTree
-normalizationTests = testGroup "normalization" [ constantFolding ]
+normalizationTests = testGroup "normalization" [ constantFolding
+                                               , fusion
+                                               ]
 
 constantFolding :: TestTree
-constantFolding = testGroup "folding of constants" [ naturalPlus ]
+constantFolding = testGroup "folding of constants" [ naturalPlus, optionalFold, optionalBuild ]
 
 naturalPlus :: TestTree
-naturalPlus = testCase "natural plus" $ normalize' (NaturalPlus (NaturalLit 1) (NaturalLit 2)) @?= NaturalLit 3
+naturalPlus = testCase "natural plus" $ do
+  isNormalized e @?= False
+  normalize' e @?= NaturalLit 3
+  where e = NaturalPlus (NaturalLit 1) (NaturalLit 2)
 
-normalize' :: Expr () () -> Expr () ()
+optionalFold :: TestTree
+optionalFold = testGroup "Optional/fold" [ just, nothing ]
+  where test label inp out = testCase label $ do
+             isNormalized (e inp) @?= False
+             normalize' (e inp) @?= out
+        e inp = (OptionalFold `App` Text `App` OptionalLit Text inp `App`
+                                             Natural `App` (Lam "j" Text (NaturalLit 1)) `App`
+                                                 (NaturalLit 2))
+        just = test "just" [TextLit "foo"] (NaturalLit 1)
+        nothing = test "nothing" [] (NaturalLit 2)
+
+optionalBuild :: TestTree
+optionalBuild = testCase "Optional/build" $ do
```

Comment:
There are two test cases worth adding to deal with common ways to implement build incorrectly.  First, you want to test that shadowing is handled correctly:

```haskell
Optional/build
Integer
(λ(optional : Type) → λ(x : Integer → optional) → λ(x : optional) → x@1 3)
```

... and you also want to test that it correctly handles an irreducible `build` like this:

```haskell
  λ(id : ∀(a : Type) → a → a)
→ Optional/build
  Integer
  ( λ(optional : type)
  → λ(just : Integer → optional)
  → λ(nothing : optional)
  → id optional (just 3)
  )
```

... where "correctly handles" means that it doesn't fail with an `error` and doesn't accidentally interpret it as a reducible `build`